packages/devtools/test/integration_tests/server_test.dart

```diff
@@ -42,34 +59,180 @@ void main() {
+    expect(serverResponse, isNotNull);
+    expect(serverResponse['clients'], hasLength(1));
+    expect(serverResponse['clients'][0]['hasConnection'], isTrue);
+    expect(serverResponse['clients'][0]['vmServiceUri'],
+        equals(appFixture.serviceUri.toString()));
+    // TODO(dantup): This test will fail until the devtools pubspec.yaml
+    // references a version of devtools_server that has this support!
+  }, timeout: const Timeout.factor(10), skip: true);
+
+  test('DevTools reports disconnects from a VM', () async {
+    // Register the VM.
+    await _send('vm.register', {'uri': appFixture.serviceUri.toString()});
+
+    // Send a request to launch DevTools in a browser.
+    await appFixture.serviceConnection
+        .callMethod(registeredServices['launchDevTools'], args: {});
+
+    // Wait for the DevTools to inform server that it's connected.
+    await _waitForClients(requiredConnectionState: true);
+
+    // Terminate the VM.
+    await appFixture.teardown();
+
+    // Ensure the client is marked as disconnected.
+    final serverResponse =
+        await _waitForClients(requiredConnectionState: false);
+    expect(serverResponse['clients'], hasLength(1));
+    expect(serverResponse['clients'][0]['hasConnection'], isFalse);
+    expect(serverResponse['clients'][0]['vmServiceUri'], isNull);
+    // TODO(dantup): This test will fail until the devtools pubspec.yaml
+    // references a version of devtools_server that has this support!
+  }, timeout: const Timeout.factor(10), skip: true);
+
+  test('server removes clients that disconnect from the API', () async {
+    // TODO(dantup): This requires the ability for us to shut down Chrome,
+    // probably via a command to the server, which needs
+    // https://github.com/dart-lang/browser_launcher/pull/12
+  }, timeout: const Timeout.factor(10), skip: true);
+
+  test('Server reuses DevTools instance if already connected to same VM',
+      () async {
+    // Register the VM.
+    await _send('vm.register', {'uri': appFixture.serviceUri.toString()});
+
+    // Send a request to launch DevTools in a browser.
+    await appFixture.serviceConnection
+        .callMethod(registeredServices['launchDevTools'], args: {});
+
+    {
+      final serverResponse =
+          await _waitForClients(requiredConnectionState: true);
+      expect(serverResponse['clients'], hasLength(1));
+    }
 
-    await appFixture.serviceConnection.streamListen(serviceName);
+    // Request again, allowing reuse, and ensure that after a few seconds there
+    // was not a second client added.
+    await appFixture.serviceConnection.callMethod(
+        registeredServices['launchDevTools'],
+        args: {'reuseWindows': true});
+
+    // This delay needs to be long enough to allow DevTools to load and connect
+    // back. If it's too short, the test may pass only because we didn't wait
+    // long enough.
+    await Future.delayed(const Duration(seconds: 5));
```

Comment:
this timeout scares me. We could get flakes on slow machines and we will be slowing down everyone's test runs by 5 seconds on fast machines. Can we instead wait for an event from the client that devtools is running?

Reply:
It wouldn't be flaky, as it'd "fail" green if the time was too short, but I agree this feels weird.

I'll add a notification each time a client connects, and we can wait on that and then ensure there's still only one connected client.

Reply:
I've added a new event (`client.launch`) that the server emits whenever a client is launched now - it includes booleans for whether it reused an existing window and/or showed a notification. For now this is only used by the tests, but it may be handy for the editors to  consume in future.